### PR TITLE
Backward compat tests

### PR DESCRIPTION
**Pull request by fulmicoton**
This PR does several things regarding backward compatibility tests:
- everything is moved to quickwit-metastore as opposed to being in an independant crate.
- the code is factorized for the three different structure we check today.
- the logic that was within the build.rs file is now integrated into the unit tests. It is a bit clunky as unit tests are modifying our test data files, but it has the merit of being simpler and making it possible to spot the lack of update in the CI
- the expected.json files get updated automatically
- the lagging expected.json files are updated.
- demux_enabled is skipped upon serialization by default.
- in test mode, the default create/update timestamps are constants.
- the backward compatibility unit tests are now regular unit tests as opposed to integration tests. (Integrations tests makes code re-use clunky)

### CODE REVIEW

File: quickwit-metastore/src/backward_compatibility_tests/README.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# Backward compatibility test project.
+
+This is just a project used to test backward compatibility of Quickwit.
+Right now, SplitMetadata, IndexMetadata and FileBasedMetadata are tested.
```
Comment (guilload) [inline quickwit-metastore/src/backward_compatibility_tests/README.md:4]:
```suggestion
Right now, `SplitMetadata`, `IndexMetadata`, and FileBackedIndex` are tested.
```

File: quickwit-metastore/src/backward_compatibility_tests/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+We want to be able to read all past versions of these files, but only write the most recent format.
+The tests consists in pairs of json file.
```
Comment (guilload) [inline quickwit-metastore/src/backward_compatibility_tests/README.md:7]:
```suggestion
The tests consist of pairs of JSON files.
```

File: quickwit-metastore/src/backward_compatibility_tests/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+XXXX.json and XXXX.expected.json
+
+XXXX.json consists in a json file in an old format.
```
Comment (guilload) [inline quickwit-metastore/src/backward_compatibility_tests/README.md:10]:
```suggestion
XXXX.json consists of a JSON file in an old format.
```

File: quickwit-metastore/src/backward_compatibility_tests/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+XXXX.expected.json consists in the expected result of
```
Comment (guilload) [inline quickwit-metastore/src/backward_compatibility_tests/README.md:11]:
```suggestion
XXXX.expected.json consists of the expected result of
```

File: quickwit-metastore/src/backward_compatibility_tests/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+serialize_new_version(deserialize(XXXX.json)).
+
+## Change of format autodetection
+
+Two things need to happen upon a change of format
```
Comment (guilload) [inline quickwit-metastore/src/backward_compatibility_tests/README.md:16]:
```suggestion
Two things need to happen upon a change of format.
```

File: quickwit-metastore/src/backward_compatibility_tests/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+### Updating expected.json
+
+Of course to make all of this work, we need to keep `*.expected.json` files up-to-date
+with the changes of format.
+
+This is done in a semi-automatic fashion.
+
+On a first pass, `deserialize(original_json) == deserialize(expectation_json)`.
+On a second pass, the tests are checking that `expectation_json = serialize(derialize(expectation_json))`.
```
Comment (guilload) [inline quickwit-metastore/src/backward_compatibility_tests/README.md:26]:
```suggestion
On a second pass, the tests are checking that `expectation_json = serialize(deserialize(expectation_json))`.
```

File: quickwit-metastore/src/backward_compatibility_tests/README.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+The two files will be identical. This is expected as this is a unit test for the
+most recent version.
+
+The unit test will start making sense in future updates thanks to the update phase
+described in the previous section.
+
+## How do I update `{IndexMetadata, SplitMetadata, FileBackedIndex}`?
+
+There two ways to types of upgrade.
```
Comment (guilload) [inline quickwit-metastore/src/backward_compatibility_tests/README.md:51]:
```suggestion
There are two types of upgrades.
```

File: quickwit-metastore/src/backward_compatibility_tests/index_metadata.rs [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    );
+    assert_eq!(
+        index_metadata.indexing_settings,
+        expected_index_metadata.indexing_settings
+    );
+    assert_eq!(
+        index_metadata.search_settings,
+        expected_index_metadata.search_settings
+    );
+    assert_eq!(
```
Comment (guilload) [inline quickwit-metastore/src/backward_compatibility_tests/index_metadata.rs:71]: FYI, after rebasing, this line will become:
```rust
assert_eq!(
        index_metadata.sources,
        expected_index_metadata.sources
    );
```

**Review by guilload — APPROVED** [review]

File: quickwit-metastore/src/backward_compatibility_tests/index_metadata.rs (hunk 8 above)
Comment (fulmicoton) [reply quickwit-metastore/src/backward_compatibility_tests/index_metadata.rs]: fixed

**Pull request merged**